sdk/go/pulumi/provider.go

```diff
@@ -141,10 +141,9 @@ func construct(ctx context.Context, req *pulumirpc.ConstructRequest, engineConn
 		return nil, err
 	}
 
-	// Wait for async work to finish.
-	if err = pulumiCtx.wait(); err != nil {
-		return nil, err
-	}
+	// Wait for async work to finish, but don't close the context until after the output properties have resolved.
+	// Note that output properties may or may not be attached to the context's waitgroup.
+	pulumiCtx.wait()
```

Comment:
> The first drain is to avoid a potential breaking change.

I'm not sure I follow why we can't just `close` after marshalling all the inputs instead of before?

Reply:
I felt it safer to drain at least once before marshalling the inputs, in case an existing provider was relying on a side effect to compute an output property. Of course, the property shouldn't resolve too eagerly but the provider might be buggy. If others think that overly cautious, am happy to remove the `wait` function.

Reply:
> in case an existing provider was relying on a side effect to compute an output property

Probably not something we should be encouraging. Even if this drain then drain again catches some of those cases it probably won't catch all async'y things. Better to not even try and let these things fail fast probably.

Reply:
I simplified it as you suggested.